Cleanup code in COPY, to handle partitions

hlinnaka opened the pull request.

melanieplageman commented on src/backend/commands/copy.c:
```diff
@@ -3687,20 +3684,12 @@ CopyFrom(CopyState cstate)
 		 * Variables for cdbpolicy
+		 *
+		 * In QE, the dispatcher sent these as part of the CopyStmt, and it
+		 * was already copied into the EState earlier.
```
I don't understand how the else could be removed and that the comment from the else would be true in the if

melanieplageman commented on src/backend/commands/copy.c:
```diff
@@ -7362,98 +7248,44 @@ FreeDistributionData(GpDistributionData *distData)
-			part_hash = d->cdbHash;
-		}
-		else
+	if (mainDistData->hashmap)
```
Do you have any idea what this comment was meant to say?

hlinnaka replied:
Hmm. I guess something like "If we are copying into a partitioned table whose partitions have differing distribution policies, get the policy for this particular child partition". I'll fix that..

Note that we don't actually support the case that partitions have different DISTRIBUTED BY clauses, e.g. one is partitioned by column A, and another by column B. However, the attribute numbers for the key columns might still be different in different partitions, if some partitions have dropped columns.

hlinnaka replied:
I added that explanation to the comment, and also added a test case for that scenario.

hlinnaka commented on src/backend/commands/copy.c:
```diff
@@ -3687,20 +3684,12 @@ CopyFrom(CopyState cstate)
 	{
 		/*
 		 * Variables for cdbpolicy
+		 *
```
You're right, that's confusing. I added a commit to this PR, as the first commit, to clarify this situation. Please take a look.

melanieplageman commented on src/backend/commands/copy.c:
```diff
@@ -7465,35 +7286,38 @@ GetDistributionPolicyForPartition(CopyState cstate, EState *estate,
+	AttrNumber	p_nattrs;	/* num of attributes in the distribution policy */
 
+	/*
+	 * These might be NULL, if we're called with a "main" GpDistributionData,
```
So does a partition table have one "main" GpDistributionData which conforms to the description in the comment below describing what a `GpDistributionData` is:
"For a partitioned table, 'policy' and 'cdbHash' are NULL. Instead, 'hashmap' contains hash table of GpDistributionData objects, for each partition."
and then the hashmap's `GpDistributionData` objects for the child partitions actually conform to the non-partitioned table version of that struct mentioned in the comment:
"For a regular, non-partitioned table, 'policy' and 'cdbHash' are filled in, and 'relid' and 'hashmap' are unused." ?

hlinnaka replied:
Yes, that's right. I reworded that comment, I hope it's more clear now.

hlinnaka commented:
On 29/11/2018 04:21, Melanie wrote:
> *@melanieplageman* approved this pull request.
> 
> LGTM and works for me.
> Since this is a refactor, I don't think that a test is required. However 
> when I was trying out the code, I checked out what tests we have for 
> COPY for partition tables and I see some in 
> src/test/regress/sql/partition.sql
> I tried COPY FROM with the multi-level partition table region in that 
> file (it worked, of course), but it got me thinking that maybe it would 
> be valuable to have a COPY test case with a partition table with dropped 
> columns/some other situation leading to attribute numbers differing 
> across partitions. This is not a test that would fail without this code, 
> but it might be a nice test to have.

I added a little debugging elog(NOTICE) in that codepath, and ran the 
regression tests, to see if we have any tests for it already. It turned 
up this, in 'alter_distribution_policy' test:

-- Check divergent distribution policies for partitioning.
create table atsdb (i int, j int, k int) distributed by (i) partition by 
range(k)
(start(1) end(4) every(1));
alter table atsdb_1_prt_2 set distributed by (j); -- this fails
alter table atsdb_1_prt_3 set distributed randomly;
-- test COPY
copy atsdb from stdin delimiter '|';

So we do have some coverage of that code. That's a different scenario 
than what I envisioned, though. There are no dropped columns here, but a 
partition that's randomly distributed.

> So, I tried to come up with a test like that with that region table and 
> couldn't actually figure out how to have a partition table with dropped 
> columns that lead to /different/ attribute numbers across the different 
> child partitions. So, my question is, what does this test case look 
> like? And, would it be valuable test coverage?
I added a test case for the dropped column scenario, too, just in case. 
See 
https://github.com/greenplum-db/gpdb/commit/3ce663a2c4f9bc5ff2807b897f6947e8cbf01100.

- Heikki
